### ISSUE TITLE
Right arrow key does not move the cursor

### ISSUE THREAD

**Issue by nateify** [chrisant996/clink#498]
I seem to be having the same issue as #168 while on Clink 1.5.6
Output of clink info:
```
version  : 1.5.6.fc63d3
session  : 2844
injected : clink_dll_x64.dll
binaries : C:\Program Files (x86)\clink
state    : C:\Users\User\AppData\Local\clink
log      : C:\Users\User\AppData\Local\clink\clink.log
settings : C:\Users\User\AppData\Local\clink\clink_settings
history  : C:\Users\User\AppData\Local\clink\clink_history
scripts  : C:\Program Files (x86)\clink ; C:\Users\User\AppData\Local\clink
inputrc  : %clink_inputrc%
             (unset)
         : state directory
             C:\Users\User\AppData\Local\clink\.inputrc
             C:\Users\User\AppData\Local\clink\_inputrc
         : %userprofile%
             C:\Users\User\.inputrc   (LOAD)
             C:\Users\User\_inputrc
         : %localappdata%
             C:\Users\User\AppData\Local\.inputrc
             C:\Users\User\AppData\Local\_inputrc
         : %appdata%
             C:\Users\User\AppData\Roaming\.inputrc
             C:\Users\User\AppData\Roaming\_inputrc
         : %home%
             C:\Users\User\.inputrc   (exists)
             C:\Users\User\_inputrc
```

Output of clink set:
```
autosuggest.async                 True
autosuggest.enable                False
autosuggest.original_case         True
autosuggest.strategy              match_prev_cmd history completion
clink.autostart                   
clink.autoupdate                  check
clink.colorize_input              False
clink.default_bindings            bash
clink.logo                        full
clink.max_input_rows              0
clink.paste_crlf                  crlf
clink.path                        
clink.promptfilter                True
clink.update_interval             5
cmd.admin_title_prefix            
cmd.altf4_exits                   True
cmd.auto_answer                   off
cmd.ctrld_exits                   True
cmd.get_errorlevel                True
color.arg                         bold
color.arginfo                     yellow
color.argmatcher                  
color.cmd                         bold
color.cmdredir                    bold
color.cmdsep                      bold
color.comment_row                 bright white on cyan
color.description                 bright cyan
color.doskey                      bold cyan
color.executable                  
color.filtered                    bold
color.flag                        default
color.hidden                      
color.histexpand                  
color.horizscroll                 
color.input                       
color.interact                    bold
color.message                     default
color.modmark                     
color.popup                       
color.popup_desc                  
color.prompt                      
color.readonly                    
color.selected_completion         
color.selection                   
color.suggestion                  bright black
color.unexpected                  default
color.unrecognized                
debug.log_terminal                False
directories.dupe_mode             add
doskey.enhanced                   True
exec.aliases                      True
exec.commands                     True
exec.cwd                          True
exec.dirs                         True
exec.enable                       True
exec.files                        False
exec.path                         True
exec.space_prefix                 True
files.hidden                      True
files.system                      False
history.auto_expand               True
history.dont_add_to_history_cmds  exit history
history.dupe_mode                 erase_prev
history.expand_mode               not_quoted
history.ignore_space              True
history.max_lines                 25000
history.save                      True
history.shared                    False
history.show_preview              True
history.sticky_search             False
history.time_format               %F %T  
history.time_stamp                off
lua.break_on_error                False
lua.break_on_traceback            False
lua.debug                         False
lua.path                          
lua.reload_scripts                False
lua.strict                        True
lua.traceback_on_error            False
match.expand_abbrev               True
match.expand_envvars              False
match.fit_columns                 True
match.ignore_accent               True
match.ignore_case                 relaxed
match.max_fitted_matches          0
match.max_rows                    0
match.preview_rows                5
match.sort_dirs                   with
match.substring                   False
match.translate_slashes           system
match.wild                        True
prompt.async                      True
prompt.transient                  off
readline.hide_stderr              False
terminal.adjust_cursor_style      True
terminal.color_emoji              auto
terminal.differentiate_keys       False
terminal.east_asian_ambiguous     auto
terminal.emulation                auto
terminal.mouse_input              auto
terminal.mouse_modifier           
terminal.raw_esc                  False
terminal.use_altgr_substitute     False
```

I put the following in my inputrc but it didn't seem to resolve the issue:
```
"\e[C": win-cursor-forward
```

**Comment by nateify** [comment]
> I think you're saying that the Right arrow key seems to have no effect, but the Left arrow key works.
> I'll assume you can type hello then Home, and then pressing Right somehow isn't moving the cursor and it stays at the beginning of the line.
This is correct.

Here is the result of `clink echo` and pressing right then CTRL-C to close:
```
Type a key to see its key sequence string.
Press Ctrl+C when finished.
"\e[C"
"\C-C"
```

The behavior occurs when using the Clink shell shortcut in the start menu, specifically the shortcut points to: `"C:\Program Files (x86)\clink\clink.bat" --profile ~\clink`

Here is the result of pressing Alt-Shift-/ and then Right:
`Right : win-cursor-forward -- Move cursor forward, or at end of line copy character from previous command, or insert
suggestion`

**Comment by chrisant996** [comment]
@nateify Thanks for sharing info that shows Clink is receiving the <kbd>Right</kbd> correctly, and is configured to invoke `win-cursor-forward`.  That's helpful, and also mysterious.

I'm not able to reproduce any problems with `win-cursor-forward`, and I use `win-cursor-forward` quite a lot.

- Can you please describe specifically what you're doing, what you expect to happen, and what happens instead?
- How long has `win-cursor-forward` not worked for you?  Did it work before, and has stopped immediately upon upgrading to v1.5.6?  What version was it working in before?
- Are any other keys not working as well?
- What terminal program are you using?  The old default conhost, Windows Terminal, ConEmu, Cmder (which uses ConEmu), ConsoleZ, Hyper Terminal, Wez Term, or etc?  If you're not sure which, can you send a screenshot?
- What version of Windows are you using?

**Comment by nateify** [comment]
>     * Can you please describe specifically what you're doing, what you expect to happen, and what happens instead?
When I type a command and press left this works just fine but I can no longer press right to move the cursor forward.
>     * How long has `win-cursor-forward` not worked for you?  Did it work before, and has stopped immediately upon upgrading to v1.5.6?  What version was it working in before?
I upgraded from 1.5.0 to 1.5.6 not too long ago but I am not 100% sure if it was showing the issue on 1.5.6, I can try to downgrade and see.
>     * Are any other keys not working as well?
As far as I know this is the only one that is not working
>     * What terminal program are you using?  The old default conhost, Windows Terminal, ConEmu, Cmder (which uses ConEmu), ConsoleZ, Hyper Terminal, Wez Term, or etc?  If you're not sure which, can you send a screenshot?
The same behavior happens on conhost, Windows Terminal, and ConEmu.
>     * What version of Windows are you using?
Windows 10, specific build  number 10.0.19045.3086

**Comment by chrisant996** [comment]
@nateify Regarding this:

> When I type a command and press left this works just fine but I can no longer press right to move the cursor forward.

When the only explanation is "right no longer moves the cursor forward", it doesn't give me anything specific to troubleshoot.  The only thing I can say is that <kbd>Right</kbd> works for me in a fresh Clink install using the settings you've shared.  But there's no way for me to guess whether we're doing the same things, or what else might be involved.

- Please describe a specific example, and please be very specific about the steps being performed.
- Can you try binding the <kbd>Right</kbd> to some other command?  That might help narrow down whether the symptom is related to the <kbd>Right</kbd> key or to the `win-cursor-forward` command.
- Can you make a short screen recording of what you're doing, and share it?  Sometimes screen recordings reveal important clues.

**Comment by chrisant996** [comment]
@vii1 I believe you, and thank you for reporting this.

I applied the Clink settings you shared, but I can't reproduce it, and I can't find anything that seems potentially relevant in the changes from v1.5.3 through v1.5.6.

- What keyboard language are you using?  (Look in the Settings app, Time & Language, Language)
- What system codepage are you using?  (Run `chcp` to find it)

Thanks for the `clink info` output; it shows you don't have any .inputrc file, so that rules out any .inputrc configuration from being relevant.

This is very interesting, and obviously something is wrong somewhere, but I can't figure out how to reproduce it, or what it could be, or where to look.

Would you be able to capture a Time Travel Debugging trace?

To capture a Time Travel Debugging trace, you would do the following:
1. Download and install the [WinDbg Windows debugger](https://learn.microsoft.com/en-us//windows-hardware/drivers/debugger/).
2. Start cmd with Clink.
3. Follow the steps for [Attach to a process](https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/time-travel-debugging-record#attach-to-a-process) and attach to the cmd.exe process.
4. Switch to the cmd window and reproduce the problem (for example maybe <kbd>Up</kbd>, <kbd>Left</kbd>, <kbd>Right</kbd>).
5. Switch to the tracer window and end the trace.
6. Compress the resulting `.run` file into a `.zip` file.
7. Upload the `.zip` file to a file sharing system (for example OneDrive or Google Drive or whatever you use).
8. Send to me a link to the file, at the email address in my github profile (the sparrowhawk996 email address at gmail.com).
9. It will help save me time if you can list the specific keys you pressed, so I know what to expect in the Time Travel Trace.

A Time Travel Debugging trace should make it crystal clear what's happening.

Thanks!

**Comment by chrisant996** [comment]
@nateify @vii1 I think this is likely related to a display optimization that was introduced in v1.5.3:

- I think it must have a case where it accidentally misses redrawing the input line.
- I think the <kbd>Right</kbd> key _**is**_ invoking `win-cursor-forward`, but the input line isn't getting redrawn, and so the cursor stays where it was, _**looking**_ like nothing happened because the screen didn't get updated to reflect what happened.
- I think if you press <kbd>Right</kbd> a few times and then e.g. <kbd>X</kbd> that the cursor will suddenly appear in the correct place and the `X` will get inserted at the correct position.

I'm working on tracking down what's happening on your computers that could cause the display optimization to malfunction.  I'm still not able to reproduce the issue on my computer, even using the settings you've shared, so there must be something additional involved beyond just settings.

If either of you can share a Time Travel Debugging trace as described [in my previous reply](https://github.com/chrisant996/clink/issues/498#issuecomment-1741287000), that could help me track it down more quickly.

**Comment by nateify** [comment]
> @nateify @vii1 I think there is chance this could be related to a display optimization that was introduced in v1.5.3:
> 
>     * Maybe it has a case where it accidentally misses redrawing the input line.
> 
>     * Maybe the Right key _**is**_ invoking `win-cursor-forward`, but the input line isn't getting redrawn, and so the cursor stays where it was _**looking**_ like nothing happened because the screen didn't get updated to reflect what happened.
> 
>     * Maybe if you press Right a few times and then e.g. X that the cursor will suddenly appear in the correct place and the `X` will get inserted at the correct position.
> 
> 
> But, even the when display optimization skips redrawing the input line, it still explicitly updates the cursor position. So maybe the display optimization isn't involved. But I can't find anything else since v.1.4.28 that seems like it could possibly be related.
> 
> I'm working on tracking down what could cause the display optimization to malfunction. I'm still not able to reproduce the issue on my computer, even using the settings you've shared, so whatever is happening, there must be something additional involved beyond just settings.
> 
> If either of you can share a Time Travel Debugging trace as described [in my previous reply](https://github.com/chrisant996/clink/issues/498#issuecomment-1741287000), that would be enormously helpful.

I sent you a time travel trace over to your gmail.

I appreciate your diligence in working on this issue!

**Comment by nateify** [comment]
I have just installed all versions from 1.5.0 to 1.5.6 in sequence and I confirmed that the issue appears on 1.5.4. Immediately downgrading back to 1.5.3 with the same config resolves the issue and the right arrow moves the cursor forward as expected.

**Comment by chrisant996** [comment]
> I sent you a time travel trace over to your gmail.

@nateify Thank you for sharing the Time Travel trace!  It was both illuminating and embarrassing.

**Problem 1:**  Recently I got clever and started downloading the output people share from `clink set` and saving it locally as `clink_settings`.  Seems like a very convenient way to exactly duplicate a whole configuration and test it, right?  😎

It would be, except that `clink_settings` and `clink set` use slightly different formats.  😱
- `clink_settings` writes and reads `name = value`
- `clink set` reports `name        value`

✅ Ok, I've relaxed the settings loader so if there's no `=` sign then it falls back to using the first `{space}` as the delimiter.
✅ And now I can easily reproduce the original reported problem about <kbd>Right</kbd> not working.

**Problem 2:**  The problem with <kbd>Right</kbd> was introduced by 9a1a144dd145b198242a3ee4615085c0091c5596.  It has an accidental side effect when `autosuggest.enable` is `False`.

- The `forward-char` and `win-cursor-forward` commands asked autosuggest a question.  But they did it even when autosuggest was disabled.  It's always been that way.
- Now cycling or searching through history tells autosuggest to not make a suggestion.  But they tell it that even when autosuggest is disabled.  That wouldn't be a problem, except that `forward-char` and `win-cursor-forward` are asking autosuggest a question even it's disabled.

Both of those are slightly incorrect, but either one by itself isn't a problem.  Put both of them together, and then once you cycle or search history, then <kbd>Right</kbd> stops working until you use <kbd>Enter</kbd> to accept the input line and send it to CMD, and start a new input line.

✅ **This is easy to fix.**

Thank you for all the diagnostic help, and sorry for the inconvenience.

I'll publish a fix shortly; I've been waiting to publish other recent enhancements until we got this one figure out.
